Documenting the rationale behind complate

joyheron opened the pull request:
This attempts to document the rationale behind complate. https://trello.com/c/UTPUHKez/6-background-documentation-of-rationale

It is not yet integrated or linked to the main index page, but I thought it was better to get the content down somewhere now and restructure later.

Commit:
Documenting the rationale behind complate

FND commented on content/rationale.md:
```diff
@@ -0,0 +1,85 @@
+Why complate?
+=============
+
+## Abstraction
```
FWIW, I prefer underlining for both H1 and H2. I also add two leading blank lines to all headings for salience. Not sure whether my opinion actually matters here though; I probably won't be the one maintaining this.

joyheron replied:
I just learned today that underline works for h1 and h2. What do you mean by two leading blank lines? Before the heading?

FND replied:
Yeah:

```
Why complate?
=============


Abstraction
-----------

When we are developing our frontend components, we can use complate markup
abstractions ("macros") as an abstraction layer for our component. …


Composition
-----------

complate allows us to write components and then explicitly compose them
together. …
```

That makes them stand out more when reading/editing the source.

FND commented on content/rationale.md:
```diff
@@ -0,0 +1,85 @@
+
+## Abstraction
+
+When we are developing our frontend components, we can use complate markup
```
another FNDism: "front-end" (throughout, same for "back{ ,-}end")

> The term “front-end” is correct when used as a compound adjective, and the term “front end” is correct when used as a noun.

https://css-tricks.com/poll-results-front-end-front-end/

FND commented on content/rationale.md:
````diff
@@ -0,0 +1,85 @@
+    {children}
+  </nav>
+}
+```
````
That example strikes me as a little too minimalistic, as it omits parameters (which also results in that somewhat alienating, unexplained `_props` argument - which, btw, I intentionally call "param[eter]s").

Here's what I tend to use:

```jsx
<Card title="Hello World">
    <p>lorem ipsum dolor sit amet</p>
    <Button label="toggle" />
</Card>
```

(though I'd probably omit `Button` for starters)

Also, in the interest of consistency: I'd use `function Navbar() { … }` (no leading space for parentheses) and four spaces for indentation (in Markdown anyway), plus `let`, semicolons and double quotes - but see maintainership note above.

joyheron replied:
Yes, that is true.

I tend to use Navbar because I had a very bad experience in a previous product where I had to modify the navbar structure and it was _painful_ to update all of the dependent projects via Copy&Paste. We put it off for at least a few weeks.

FND replied:
That's a good point: Perhaps we could start with a straightforward, self-explanatory example that doesn't rely on context/interpretation (not actually sure whether my `Card` qualifies for that tbh), but add afterwards that [the same principle applies to larger structures as well](https://en.wikipedia.org/wiki/Fractal) - a nav bar seems like a good example there; people should be able to grok that without the need for an elaborate code sample.

joyheron replied:
I'm thinking about a new example that would word well.

By the way, I don't really like your card example because it always uses an `h3` as a card level, but the card would only know the correct semantic level in context of the document. 😬

For structural macros like cards I often use no parameters.

FND replied:
I dislike the card example for the same reason. \o/ (Which is why I cheated above by omitting the heading... 😬 )

You're far better suited than me to come up with grokkable examples, but it might be worth considering using a simple [Bootstrap component](https://getbootstrap.com/docs/4.1/components) (list group, perhaps?) because they're widely known.

FND commented on content/rationale.md:
```diff
@@ -0,0 +1,85 @@
+
+This is _extremely powerful_ and solves a major pain point when writing
+maintainable frontend code. Requirements change constantly. When a new request
+comes, or the corporate design has changed yet again, I can change the markup
```
`s/request comes/requirement comes up|in/`?

FND commented on content/rationale.md:
```diff
@@ -0,0 +1,85 @@
+## Abstraction
+
+When we are developing our frontend components, we can use complate markup
+abstractions ("macros") as an abstraction layer for our component. Consider
```
`s/abstraction layer/encapsulation layer/` perhaps? Mentioning encapsulation seems important (see below), plus it would avoid redundant "abstraction".

FND commented on content/rationale.md:
```diff
@@ -0,0 +1,85 @@
+maintainable frontend code. Requirements change constantly. When a new request
+comes, or the corporate design has changed yet again, I can change the markup
+of my component _without breaking any part of my application that uses the
+component_.
```
That seems like a potentially misleading promise, as various other parts might rely on that markup (CSS, JavaScript, tests). Perhaps we could argue with encapsulation instead: Change the definition and all instances are updated automatically? I might be overly cautious though.

More generally, I'd argue with less specific scenarios - in particular, external/corporate requirements being the defining factor kinda offends my anti-authoritarian/-capitalist proclivities; in my experience, markup adjustments are often intrinsically motivated. (Obviously that, too, is an FNDism, so YMMV.)

FND commented on content/rationale.md:
```diff
@@ -0,0 +1,85 @@
+
+Another major benefit is that it is possible to write complex frontend
+components and others can _use_ the components easily, without having to
+understand how the component works under the hood.
```
I'm always careful to add that I'm not trying to shield users from the underlying markup; they should still be aware of that markup and know why it matters.

joyheron replied:
Good point.

I just know that some people on my team love complate, because they can use existing components and make pretty UIs without actually having to write the CSS themselves.

FND replied:
Yeah, I know, same here. But deep inside, I remain a naive, quixotic idealist...

FND commented on content/rationale.md:
```diff
@@ -0,0 +1,85 @@
+complate allows us to write components and then explicitly compose them
+together. For instance, when I write a navbar component, I can reuse the
+component in multiple projects and change the content of the navbar without
+affecting it.
```
I know what you mean, but "change the content without affecting it" sounds confusing - perhaps something along the lines of "without worrying about the surrounding boilerplate structures"?

FND commented on content/rationale.md:
````diff
@@ -0,0 +1,85 @@
+
+```jsx
+function NavbarItem ({ href, current }, ...children) {
+	const className = current ? 'currentPage' : ''
````
FYI: `null` instead `''` of would omit the `class` attribute entirely. (Also, CamelCase in class names is a crime. 🙂 )

joyheron replied:
should I use `class={current && 'current'}`

joyheron replied:
I just think that `const className = current ? 'current' : null` doesn't look so nice (I don't like explicitly using null). Is there a better way? Could you use `className = current && 'current'` ? Although I don't know if that would be more readable.

FND replied:
Nothing wrong with explicit `null` IMO. 🤷‍♂

`&&` works and is probably idiomatic in the JSX/React world, but I agree that it might be confusing in this context (it's not necessarily obvious that JS's logical operators work this way - plus it might be brittle; think `0`).

joyheron replied:
I think `className = current ? 'current' : null` is definitely easier to read.

FND reviewed (commented):
Excellent job, thank you very much!

I just have a few persnickety notes and suggestions; feel free to ignore them.

Commit:
Some fixes to rationale, still thinking about more improvements.

Commit:
More fixes to rationale

joyheron commented:
I've done some revision to try to correct the mentioned items.

FND committed:
tweaked rationale

minor phrasing details as well as consistency adjustments (whitespace,
quotation marks in code)

FND commented:
Looking good! I've tweaked a few minor details in 7ad93a84d3875e406772ad6f606affa820f35619 - feel free to cherry-pick or disregard at will.

At some point we should consider adding some sort of current-affairs contextualization: For people used to writing SPAs (and React in particular, of course), none of this is new - so it's not immediately obvious how complate is different (SSR is only mentioned as an aside towards the bottom). However, that shouldn't stop us from merging this now.

moonglum commented:
Excellent explanation 👏

joyheron committed:
Merge pull request #2 from complate/rationale-tweaks

tweaked rationale

joyheron commented:
@FND Thank you for the feedback. I've integrated your feedback and we could merge this as far as I am concerned. :)

FND commented:
Merged as 1873d9b30cd0c2f254b630fc6b6a64525ff87e23 ff. (I've merely just linearized the history) - thank you!

https://complate.github.io/rationale.html
